omod/src/main/java/org/openmrs/module/smartonfhir/web/servlet/SmartAccessConfirmation.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public License,
+ * v. 2.0. If a copy of the MPL was not distributed with this file, You can
+ * obtain one at http://mozilla.org/MPL/2.0/. OpenMRS is also distributed under
+ * the terms of the Healthcare Disclaimer located at http://openmrs.org/license.
+ *
+ * Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
+ * graphic logo is a trademark of OpenMRS Inc.
+ */
+package org.openmrs.module.smartonfhir.web.servlet;
+
+import javax.crypto.spec.SecretKeySpec;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import java.io.IOException;
+import java.net.URLDecoder;
+import java.net.URLEncoder;
+import java.nio.charset.StandardCharsets;
+
+import org.keycloak.crypto.Algorithm;
+import org.keycloak.crypto.JavaAlgorithm;
+import org.keycloak.crypto.KeyWrapper;
+import org.keycloak.crypto.MacSignatureSignerContext;
+import org.keycloak.crypto.SignatureSignerContext;
+import org.keycloak.jose.jws.JWSBuilder;
+import org.keycloak.representations.JsonWebToken;
+import org.openmrs.User;
+import org.openmrs.api.context.Context;
+import org.openmrs.module.smartonfhir.util.SmartSecretKeyHolder;
+
+public class SmartAccessConfirmation extends HttpServlet {
+	
+	public void doGet(HttpServletRequest req, HttpServletResponse res) throws IOException {
+		String token = req.getParameter("token");
+		
+		if (token == null) {
+			res.sendError(HttpServletResponse.SC_BAD_REQUEST);
+			return;
+		}
+		JsonWebToken tokenSentBack = new JsonWebToken();
+		User user = Context.getAuthenticatedUser();
+		if (user == null) {
+			res.sendError(HttpServletResponse.SC_UNAUTHORIZED);
```

Comment:
So, to make this work properly, we need to redirect the user back to Keycloak, but without a token value, and then inside the Keycloak Authentication provider run something like `context.attempted()` to indicate that the attempt to login this way was unsuccessful. Merely returning a 401 error here will just keep the user on a "Unauthorized" page of some sort.

Reply:
Most likely this looks something like:

```java
res.sendRedirect(decodedUrl.replace("{APP_TOKEN}", ""));
```

Reply:
@ibacher thanks I will change this.